src/lib.rs

```diff
@@ -170,21 +170,33 @@ impl<A: Discard> DiscardOnDrop<A> {
     /// This is implemented as a function (*not* a method) so that way it doesn't interfere with any of the
     /// methods on `discarder`.
     #[inline]
-    pub fn leak(mut this: Self) -> A {
-        match this.0.take() {
-            Some(value) => value,
-            None => unreachable!(),
+    pub fn leak(this: Self) -> A {
+        // We want to move the `A` out of `this`, but that's not allowed because `this` implements `Drop`
+        // (and we must also avoid calling `drop()` on `this` or else `A` would get dropped twice).
+        // We can do that move by using the unsafe function std::ptr::read(),
+        // and then use `mem::forget()` on `this` so it never gets dropped. The `A` will get dropped by the caller.
+        unsafe {
```

Comment:
There are 2 possible implementations here:

1. The one in my commit, where you read the `A` directly out of the `DiscardOnDrop<A>` using `ptr::read()`. This `A` is passed onto the caller who will then eventually drop it. The `DiscardOnDrop<A>` and the  `ManuallyDrop<A>` never get dropped.

2. An implementation where you read the `ManuallyDrop<A>` out of the `DiscardOnDrop<A>` using `ptr::read()`, then use `into_inner` to get the `A` out of the `ManuallyDrop<A>`. Such an implementation could look like this:

```rust
pub fn leak(this: Self) -> A {
    unsafe {
        let manually_drop = ::std::ptr::read(&this.0);
        ::std::mem::forget(this);
        manually_drop.into_inner()
    }
}
```

There are only some very small differences between these 2 implementations. In this second implementation, the `ManuallyDrop<A>` does get dropped instead of leaked, but `ManuallyDrop` doesn't actually do anything when dropped (it's just a `union` with 1 element). Theoretically, the Rust standard library could change the implementation of `ManuallyDrop<A>` to do allocations or read files or do other things that need to get deallocated when the `ManuallyDrop` gets dropped, and in that case we'd be leaking those resources by not dropping the `ManuallyDrop`, but in practice that's just not going to happen.

Mostly it's just a matter of taste which implementation you prefer. I can do the 2nd one if you like it more.

Reply:
I'm not worried about such hypothetical changes to the Rust stdlib.